tflint/runner.go

```diff
@@ -501,6 +502,69 @@ func (r *Runner) LookupIssues(files ...string) Issues {
 	return issues
 }
 
+// WalkExpressions visits all blocks that can contain expressions:
+// resource, data, module, provider, locals, and output. It calls the walker
+// function with every expression it encounters and halts if the walker
+// returns an error.
+func (r *Runner) WalkExpressions(walker func(hcl.Expression) error) error {
+	visit := func(expr hcl.Expression) error {
+		return r.WithExpressionContext(expr, func() error {
+			return walker(expr)
+		})
+	}
+
+	for _, resource := range r.TFConfig.Module.ManagedResources {
+		if err := r.walkBody(resource.Config.(*hclsyntax.Body), visit); err != nil {
```

Comment:
This type-casting panics when inspecting JSON syntax.

```
% tflint
Panic: interface conversion: hcl.Body is *json.body, not *hclsyntax.Body
 -> 0: main.main.func1: /main.go(30)
 -> 1: runtime.gopanic: /panic.go(967)
 -> 2: runtime.panicdottypeE: /iface.go(260)
 -> 3: runtime.panicdottypeI: /iface.go(270)
 -> 4: github.com/terraform-linters/tflint/tflint.(*Runner).WalkExpressions: /runner.go(517)
 -> 5: github.com/terraform-linters/tflint/rules/terraformrules.(*TerraformUnusedDeclarationsRule).Check: /terraform_unused_declaration.go(53)
 -> 6: github.com/terraform-linters/tflint/cmd.(*CLI).inspect: /inspect.go(57)
 -> 7: github.com/terraform-linters/tflint/cmd.(*CLI).Run: /cli.go(82)
 -> 8: main.main: /main.go(43)
 -> 9: runtime.main: /proc.go(203)
 -> 10: runtime.goexit: /asm_amd64.s(1373)
```

However, I'm not sure how to walk all expressions on `json.body`. In this case, one way is skipping the analysis.

Reply:
Makes sense. I'll look into what it might take to walk it, but the unexported type is a bad sign. Might just have to skip for now. I can imagine visiting every leaf node but that's not entirely correct since a `map` can have keys that are non-static expressions.